File: 706.design-hash-map.cs
/*
 * @lc app=leetcode id=706 lang=csharp
 *
 * [706] Design HashMap
 */

// @lc code=start
public class MyHashMap
{
    private List<(int, int)>[] arr;

    public MyHashMap()
    {
        arr = new List<(int, int)>[1 << 15];
        for (int i = 0; i < (1 << 15); i++)
        {
            arr[i] = new List<(int, int)>();
        }
    }

    public int EvalHash(int key)
    {
        return ((key * 1031231) & ((1 << 20) - 1)) >> 5;
    }

    public void Put(int key, int value)
    {
        int t = EvalHash(key);
        for (int i = 0; i < arr[t].Count; i++)
        {
            if (arr[t][i].Item1 == key)
            {
                arr[t][i] = (key, value);
                return;
            }
        }
        arr[t].Add((key, value));
    }

    public int Get(int key)
    {
        int t = EvalHash(key);
        for (int i = 0; i < arr[t].Count; i++)
        {
            if (arr[t][i].Item1 == key)
            {
                return arr[t][i].Item2;
            }
        }
        return -1;
    }

    public void Remove(int key)
    {
        int t = EvalHash(key);
        for (int i = 0; i < arr[t].Count; i++)
        {
            if (arr[t][i].Item1 == key)
            {
                arr[t].RemoveAt(i);
                return;
            }
        }
    }
}


/**
 * Your MyHashMap object will be instantiated and called as such:
 * MyHashMap obj = new MyHashMap();
 * obj.Put(key,value);
 * int param_2 = obj.Get(key);
 * obj.Remove(key);
 */
// @lc code=end

File: 706.design-hash-map.py
#
# @lc app=leetcode id=706 lang=python3
#
# [706] Design HashMap
#

# @lc code=start
class MyHashMap:
    '''
    Multiplicative hashing: a K MOD 2 ^ w / 2 ^ (w - m): a is relatively prime to 2 ^ w: we choose a = 1031231, w = 20 and m = 15 and I choose w = 20. m is chosen such that 1 << m > number of operation (potential collisions). Also we use the trick: for any s: s % (2^t) = s & (1<<t) - 1
    '''
    def eval_hash(self, key):
        return ((key * 1031231) & (1<<20) - 1) >> 5

    def __init__(self):
        self.arr = [[] for _ in range(1 << 15)]
        
    def put(self, key, value):
        t = self.eval_hash(key)
        for i,(k, v) in enumerate(self.arr[t]):
            if k == key:
                self.arr[t][i] = (k, value)
                return
        self.arr[t].append((key, value))

    def get(self, key):
        t = self.eval_hash(key)
        for i,(k, v) in enumerate(self.arr[t]):
            if k == key: 
                return v
        return -1

    def remove(self, key: int):
        t = self.eval_hash(key)
        for i,(k, v) in enumerate(self.arr[t]):
            if k == key:
                self.arr[t].remove((k,v))

# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)
# Your MyHashMap object will be instantiated and called as such:
# obj = MyHashMap()
# obj.put(key,value)
# param_2 = obj.get(key)
# obj.remove(key)
# @lc code=end

